File: Arrays_and_Hashing/Group_Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
let groupAnagrams = (strs) => {
    let result = [];
    let anagramMap = {};

    strs.forEach(s => {
        let sorted = s.split("").sort().join();
        anagramMap[sorted] ? anagramMap[sorted].push(s) : anagramMap[sorted] = [s];
    });

    return Object.values(anagramMap);
}

File: Arrays_and_Hashing/GroupAnagrams.go
func groupAnagrams(strs []string) [][]string {
    wordMap := make(map[[26]int][]string)

    for _, word := range strs {
        var letterCount [26]int
        for _, char := range word {
            letterCount[char - 'a']++
        }
        wordMap[letterCount] = append(wordMap[letterCount], word)
    }

    var result [][]string
    for _, words := range wordMap {
        result = append(result, words)
    }
    return result
}

// We need to track each word
// -> Store sorted letters from word as a key in a hash, values are words
// Make one pass to store keys, another pass to store values
// -> O(N) * O(N)
// Iterate through hashMap and build return array
// Handle empty strings
// Handle single element
